### PR TITLE
feat: Helm async deploy Devtron Apps

### PR DESCRIPTION
**Pull request by Ash-exp**
# Description

- Currently in Devtron Apps with Helm installation, we are performing the Install/Upgrade in synchronous mode. With this flow, in some cases the API context expires as the Install/ Upgrade may take too long. The common cases for this behaviour are:
  - The deployment charts have per or post hooks/ jobs/ heavy resources to build
  - The cluster connection in kubelink takes longer due to the difference in region.

- In this PR we are performing Helm install/ Upgrade for devtron Apps in asynchronous  mode. This feature will flag based. Use `ENABLE_ASYNC_INSTALL_DEVTRON_CHART` (bool) in config map: `devtron-cm` to enable/disable this feature. Default value is set to `false`.
    ```
    ENABLE_ASYNC_INSTALL_DEVTRON_CHART: "true"
    ```

- Other Environment flags are: (config map: `devtron-cm`)
  1. **gRPC Timeout For Devtron App Deployment**: `DEVTRON_CHART_INSTALL_REQUEST_TIMEOUT` (in mins) default 6

      ```
      DEVTRON_CHART_INSTALL_REQUEST_TIMEOUT: "7"
      ```
  2. **Nats Configuration**: `CONSUMER_CONFIG_JSON`(json) Default natsMsgProcessingBatchSize is 1. 
      
      ```
      CONSUMER_CONFIG_JSON: '{"DEVTRON-CHART-INSTALL-DURABLE":{"natsMsgProcessingBatchSize":2}}'
      ```

Fixes https://github.com/devtron-labs/devtron/issues/4264

## How Has This Been Tested?
Test Cases:
- [x] Single Deployment:
  - [x] GitOps: (Auto/Manual) // Sanity
    - [x] With Custom charts (enforce pending-install state)
    - [x] With Devtron Reference charts
    - [x] With last deployed config
    - [x] With last saved config
    - [x] With a specific config
  - [x] Helm: (Auto/Manual)
    - [x] With Custom charts (enforce pending-install state)
    - [x] With Devtron Reference charts
    - [x] With last deployed config
    - [x] With last saved config
    - [x] With a specific config
- [x] Env Flag verification (enable/disable)
- [x] Gitops to Helm conversion
- [x] Hibernate and Restore Deployment
- [x] Deploy externally (Cluster CLI), then deploy from UI 
- [x] App Group Bulk Deploy UI: (More than 20 application) 
- [x] Bulk Deployment API: /orchestrator/batch/v1beta1 (More than 20 application) 
- [x] Bulk Hibernate API: (More than 20 application)

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code.
* [x] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/app/AppService.go [hunk 1]
```diff
@@ -3110,19 +3151,7 @@ func (impl *AppServiceImpl) createHelmAppForCdPipeline(overrideRequest *bean.Val
...
-				// retry install
-				_, err = impl.helmInstallReleaseWithCustomChart(ctx, releaseIdentifier, referenceChartByte, mergeAndSave)
-
-				// if retry failed, return
-				if err != nil {
-					impl.logger.Errorw("release not found, failed to re-install helm application", "err", err)
-					return false, err
-				}
-			} else if err != nil {
+			if err != nil {
```
Comment (Ash-exp) [inline pkg/app/AppService.go:3154]: add comment

File: pkg/app/AppService.go [hunk 2]
```diff
@@ -1884,8 +1890,43 @@ func (impl *AppServiceImpl) GetTriggerEvent(deploymentAppType string, triggeredA
 	return triggerEvent
 }
 
-func (impl *AppServiceImpl) TriggerRelease(overrideRequest *bean.ValuesOverrideRequest, ctx context.Context, triggeredAt time.Time, deployedBy int32) (releaseNo int, manifest []byte, err error) {
-	triggerEvent := impl.GetTriggerEvent(overrideRequest.DeploymentAppType, triggeredAt, deployedBy)
+func (impl *AppServiceImpl) HandleCDTriggerRelease(overrideRequest *bean.ValuesOverrideRequest, ctx context.Context, triggeredAt time.Time, deployedBy int32) (releaseNo int, manifest []byte, err error) {
+	if impl.isDevtronAsyncInstallModeEnabled(overrideRequest.DeploymentAppType) {
+		event := &bean.AsyncCdDeployEvent{
```
Comment (Ash-exp) [inline pkg/app/AppService.go:1895]: move into a separate function

File: pkg/app/AppService.go [hunk 3]
```diff
@@ -103,6 +104,8 @@ type AppServiceConfig struct {
 	GetPipelineDeployedWithinHours      int    `env:"DEPLOY_STATUS_CRON_GET_PIPELINE_DEPLOYED_WITHIN_HOURS" envDefault:"12"` //in hours
 	HelmPipelineStatusCheckEligibleTime string `env:"HELM_PIPELINE_STATUS_CHECK_ELIGIBLE_TIME" envDefault:"120"`             //in seconds
 	ExposeCDMetrics                     bool   `env:"EXPOSE_CD_METRICS" envDefault:"false"`
+	EnableAsyncInstallDevtronChart      bool   `env:"ENABLE_ASYNC_INSTALL_DEVTRON_CHART" envDefault:"true"`
+	DevtronChartInstallTimeout          int    `env:"DEVTRON_CHART_INSTALL_TIMEOUT" envDefault:"6"` //in minutes
```
Comment (Ash-exp) [inline pkg/app/AppService.go:108]: rename to `DEVTRON_CHART_INSTALL_REQUEST_TIMEOUT`

File: pkg/app/AppService.go [hunk 4]
```diff
@@ -3162,55 +3191,77 @@ func (impl *AppServiceImpl) createHelmAppForCdPipeline(overrideRequest *bean.Val
...
+		_, err = impl.cdWorkflowRepository.SaveWorkFlowRunner(runner)
+		if err != nil {
+			impl.logger.Errorw("err on updating cd workflow runner for status update", "err", err)
+			return err
+		}
+	} else {
+		// if the current cdWfr status is already a terminal status and then don't update the status
+		// e.g: Status : Failed --> Progressing (not allowed)
+		if slices.Contains(pipelineConfig.WfrTerminalStatusList, cdWfr.Status) {
+			impl.logger.Errorw("deployment has already been terminated for workflow runner", "workflowRunnerId", cdWfr.Id, "err", err)
```
Comment (Ash-exp) [inline pkg/app/AppService.go:3247]: use log type: warning

File: pkg/pipeline/WorkflowDagExecutor.go [hunk 5]
```diff
@@ -306,6 +312,70 @@ func (impl *WorkflowDagExecutorImpl) Subscribe() error {
 	return nil
 }
 
+func (impl *WorkflowDagExecutorImpl) SubscribeDevtronAsyncInstallRequest() error {
```
Comment (Ash-exp) [inline pkg/pipeline/WorkflowDagExecutor.go:315]: update name with helm

File: internal/sql/repository/pipelineConfig/CdWorfkflowRepository.go [hunk 6]
```diff
@@ -591,14 +614,16 @@ func (impl *CdWorkflowRepositoryImpl) FetchArtifactsByCdPipelineId(pipelineId in
 
 func (impl *CdWorkflowRepositoryImpl) GetLatestTriggersOfHelmPipelinesStuckInNonTerminalStatuses(getPipelineDeployedWithinHours int) ([]*CdWorkflowRunner, error) {
 	var wfrList []*CdWorkflowRunner
+	excludedStatusList := WfrTerminalStatusList
+	excludedStatusList = append(excludedStatusList, WorkflowInitiated, WorkflowInQueue)
```
Comment (kripanshdevtron) [inline internal/sql/repository/pipelineConfig/CdWorfkflowRepository.go:618]: we should also add Starting state in this list

**Review by vikramdevtron — DISMISSED** [review]

File: internal/sql/repository/pipelineConfig/CdWorfkflowRepository.go (hunk 6 above)
Comment (Ash-exp) [reply internal/sql/repository/pipelineConfig/CdWorfkflowRepository.go]: updated 👍

**Review by vikramdevtron — APPROVED** [review]

**Pull request merged**